### PR TITLE
RDRP-781: Unit test for load_validate_mapper()

### PR DESCRIPTION
**Pull request by JenCheshire**
Ticket changed: hard coded path were relative so just unit tests for:
load_validate_mapper()

Still require test for:
load_val_snapshot_json()

Removed as no longer needed:
load_validate_secondary_snapshot()

## Pull Request submission

*Insert detailed bullet points about your changes here!*

*Insert any instructions to help the reviewer, e.g. "install new requirements from `requirements.txt`"*

*Let the reviewer know what data files are needed (and if applicable, where they are to be found)

#### Closes or fixes

* Detail the ticket(s) you are closing with this PR
Closes #

#### Code

- [ ] **Code runs** The code runs on my machine and/or CDSW
- [ ] **Conflicts resolved** There are no conflicts (I have performed a rebase if necessary)
- [ ] **Requirements** My/our code functions according to the requirements of the ticket
- [ ] **Dependencies** I have updated the environment yaml so it includes any new libraries I have used
- [ ] **Configuration file updated** any high level parameters that the user may interact with have been put into the config file (and imported to the script)
- [ ] **Clean Code**
    - [ ] Code is as [PEP 8]([url](https://peps.python.org/pep-0008/)) compliant as I can humanly make it
    - [ ] Code passess flake8 linting check
    - [ ] Code adheres to [DRY](https://en.wikipedia.org/wiki/Don%27t_repeat_yourself)
- [ ] **Type hints** All new functions have [type hints ](https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html)

#### Documentation

Any new code includes all the following forms of documentation:

- [ ] **Function Documentation** Docstrings within the function(s')/methods have been created
    - [ ] Includes `Args` and `returns` for all major functions
    - [ ] The docstring details data types
- [ ] **Updated Documentation**: User and/or developer working doc has been updated

#### Data
- [ ] All data needed to run this script is available in Dev/Test
- [ ] All data is excluded from this pull request
- [ ] Secrets checker pre-commit passes

#### Testing
- [ ] **Unit tests** Unit tests have been created and are passing _or a new ticket to create tests has been created_

---

# Peer Review Section

- [ ] All requirements install from (updated) `requirements.txt`
- [ ] Documentation has been created and is clear - **check the working document**
- [ ] Doctrings ([Google format](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)) have been created and accurately describe the function's functionality
- [ ] Unit tests pass, or if not present _a new ticket to create tests has been created_
- [ ] **Code runs** The code runs on reviewer's machine and/or CDSW

#### Final approval (post-review)

The author has responded to my review and made changes to my satisfaction.
- [ ] **I recommend merging this request.**

---

### Review comments

*Insert detailed comments here!*

These might include, but not exclusively:

- bugs that need fixing (does it work as expected? and does it work with other code
  that it is likely to interact with?)
- alternative methods (could it be written more efficiently or with more clarity?)
- documentation improvements (does the documentation reflect how the code actually works?)
- additional tests that should be implemented (do the tests effectively assure that it
  works correctly?)
- code style improvements (could the code be written more clearly?)
- Do the changes represent a change in functionality so the version number should increase? Start a discussion if so.
- As a review you can generates the same outputs from running the code

Your suggestions should be tailored to the code that you are reviewing.
Be critical and clear, but not mean. Ask questions and set actions.

### CODE REVIEW

**Comment by JenCheshire** [comment]
Hi @CBROWN-ONS and @AnneONS - I've requested a review from you both.  Really keen for any ideas to make these better.  The new tests are testing functions that contain paths to the config.  If they are run in full, they go to real configs which contain all the variables so they error on the test data.  I couldn't find a way to avoid this, however, because those functions are already unit tested, I have concentrated on checking that the appropriate functions are called within these tests.  Grateful for your thoughts.  Thanks

File: tests/test_staging/test_staging_helpers.py [hunk 1]
```diff
@@ -5,17 +5,21 @@
...
 from typing import Tuple
 from datetime import date
+import logging
 
 # Third Party Imports
+import pytest
 import pandas as pd
 import numpy as np
 import pyarrow.feather as feather
+from unittest.mock import MagicMock, patch
```
Comment (cbrowning04) [inline tests/test_staging/test_staging_helpers.py]: `unittest` is part of the standard python library

Comment (cbrowning04) [inline tests/test_staging/test_staging_helpers.py]: MagicMock unused

File: tests/test_staging/test_staging_helpers.py [hunk 2]
```diff
@@ -247,10 +298,91 @@ def test_load_snapshot_feather(self, tmp_path):
...
+        # Assertions
+        assert (
+            res_rate == expected_res_rate
+        ), "The response rate calculated by load_val_snapshot_json does not match the expected value."
+        mock_load_json.assert_called_once_with(snapshot_path)
+        mock_validate_data_with_schema.assert_called()
+        mock_combine_schemas_validate_full_df.assert_called()
+
+
+class TestLoadValSecondarySnapshotJson(object):
```
Comment (cbrowning04) [inline tests/test_staging/test_staging_helpers.py]: Are we able to change the name of this test class? This shouldn't be shorthanded and should reflect the function that it's testing

File: tests/test_staging/test_staging_helpers.py [hunk 3]
```diff
@@ -428,7 +547,6 @@ def pc_mapper_output(self) -> pd.DataFrame:
         df = pd.DataFrame(columns=columns, data=data)
         return df
 
-
     def get_todays_date(self) -> str:
         """Get the date in the format YYYY-MM-DD. Used for filenames."""
         today = date.today()
```
Comment (cbrowning04) [inline tests/test_staging/test_staging_helpers.py]: ISSUES RELEVANT TO THE CODE OUTSIDE OF THIS PR
- config param unused in `load_validate_secondary_snapshot`

File: tests/test_staging/test_staging_helpers.py [hunk 4]
```diff
@@ -247,10 +298,91 @@ def test_load_snapshot_feather(self, tmp_path):
...
+        mock_validate_data_with_schema.assert_called()
+        mock_combine_schemas_validate_full_df.assert_called()
+
+
+class TestLoadValSecondarySnapshotJson(object):
+    """Tests for the load_validate_secondary_snapshot function."""
 
+    @patch("src.utils.local_file_mods.load_local_json")
+    @patch("src.staging.validation.validate_data_with_schema")
+    @patch("src.staging.spp_snapshot_processing.full_responses")
```
Comment (cbrowning04) [inline tests/test_staging/test_staging_helpers.py]: unused

File: tests/test_staging/test_staging_helpers.py [hunk 5]
```diff
@@ -247,10 +298,91 @@ def test_load_snapshot_feather(self, tmp_path):
...
+    @patch("src.staging.validation.combine_schemas_validate_full_df")
+    def test_load_validate_secondary_snapshot(
+        self,
+        mock_combine_schemas_validate_full_df,
+        mock_full_responses,
```
Comment (cbrowning04) [inline tests/test_staging/test_staging_helpers.py]: unused

File: tests/test_staging/test_staging_helpers.py [hunk 6]
```diff
@@ -247,10 +298,91 @@ def test_load_snapshot_feather(self, tmp_path):
...
+    """Tests for the load_val_snapshot_json function."""
+
+    @patch("src.utils.local_file_mods.load_local_json")
+    @patch("src.staging.validation.validate_data_with_schema")
+    @patch("src.staging.spp_snapshot_processing.full_responses")
+    @patch("src.staging.validation.combine_schemas_validate_full_df")
+    def test_load_val_snapshot_json_correct_response_rate(
+        self,
+        mock_combine_schemas_validate_full_df,
+        mock_full_responses,
```
Comment (cbrowning04) [inline tests/test_staging/test_staging_helpers.py]: unused

File: tests/test_staging/test_staging_helpers.py [hunk 7]
```diff
@@ -247,10 +298,91 @@ def test_load_snapshot_feather(self, tmp_path):
         ), "Snapshot df has more columnss than expected."
 
 
-# load_val_snapshot_json [CANT TEST: TOO MANY HARD CODED PATHS]
+class TestLoadValSnapshotJson(object):
+    """Tests for the load_val_snapshot_json function."""
+
+    @patch("src.utils.local_file_mods.load_local_json")
+    @patch("src.staging.validation.validate_data_with_schema")
+    @patch("src.staging.spp_snapshot_processing.full_responses")
```
Comment (cbrowning04) [inline tests/test_staging/test_staging_helpers.py]: unused

**Review by cbrowning04 — DISMISSED** [review]
Hi Jen.

I've not come across much of the unittest package. so this was a nice thing for me to learn reading through your code :).

While the mocking is neat, and seems to work as intended, I have some suggestions for improving the maintainability and robustness of the tests.

To begin with, there are functions within the `load_validate_mapper` function that aren't being included in the mock tests, where as some are. This is rather inconsistent in a way.  

Next, the function is able to be tested using real data, rather than mocks. For example, you are able to write out schemas/configs/jsons to temp paths, and read them back in, using pytest. This allows for real testing of the function, ensuring it is able to act as intended, and not just call the functions. This will more accurately test that the function is acting in the correct way, making the tests more secure.

Using pytest would also align it with the rest of the tests in this module.

In general, the main processes of this function include reading from files, therefore it seems like a requirement that this is happening during the tests, rather than mocks being implemented.

More tests can also be included, such as checking that the messages are being written to the logger. This is unrelated to the mocking approach, and more of a feature request.

Thanks, I'm happy to answer any questions you might have.

File: tests/test_staging/test_staging_helpers.py (hunk 1 above)
Comment (JenCheshire) [reply tests/test_staging/test_staging_helpers.py]: Removed

**Comment by AnneONS** [comment]
HI @CBROWN-ONS, thanks for your review and writing out your thoughts clearly.
The main reason we're keen on mocking in RDSA is because it's needed when you're developing on hdfs, as we were at the beginning of this project.
I recently updated the devtest version of the code so it would be ready for transfer to the Cloud, and George is keen our project can be used as a test case for the important DAP to Cloud transfer.
Unfortunately the unit tests you created in the output modules don't work in devtest, there is a permissions error.

I could show you this if you are interested.

File: tests/test_staging/test_staging_helpers.py (hunk 2 above)
Comment (JenCheshire) [reply tests/test_staging/test_staging_helpers.py]: Done

File: tests/test_staging/test_staging_helpers.py (hunk 7 above)
Comment (JenCheshire) [reply tests/test_staging/test_staging_helpers.py]: This is being used

File: tests/test_staging/test_staging_helpers.py (hunk 6 above)
Comment (JenCheshire) [reply tests/test_staging/test_staging_helpers.py]: As above, I think the mock for the function full_responses is being used

File: tests/test_staging/test_staging_helpers.py (hunk 4 above)
Comment (JenCheshire) [reply tests/test_staging/test_staging_helpers.py]: As above

File: tests/test_staging/test_staging_helpers.py (hunk 5 above)
Comment (JenCheshire) [reply tests/test_staging/test_staging_helpers.py]: As above

**Review by AnneONS — APPROVED** [review]
Thank you so much for all your hard work on this!! Hopefully it was a good learning point as well as probably a pain point!

**Pull request merged**